### PR TITLE
Fix builds for macOS

### PR DESCRIPTION
**Pull request by eugene-babichenko**
This pull request introduces:

- Successful builds with macOS GCC (Homebrew GCC 7.4.0 & 9.1.0)
- Using `conan` to fetch dependencies instead of just using `git`
- Several modification in `CMakeLists.txt` for readability.

**WIP**

- macOS clang builds

### CODE REVIEW

**Comment by carlocayos** [comment]
Please create a PR to a `develop` branch first

**Comment by eugene-babichenko** [comment]
@carlocayos Done.

File: src/infrastructure/utils/variadic_struct.h [hunk 1]
```diff
@@ -363,9 +364,9 @@ namespace nem2_sdk { namespace internal {
 	{ };
 	
 	template<size_t I> struct struct_field_by_index<I, NullStruct> {
-		static_assert(I == 0, "variadic struct field index is out of range");
+		static_assert(I < 0, "variadic struct field index is out of range");
```
Comment (eugene-babichenko) [inline src/infrastructure/utils/variadic_struct.h:371]: Fixed after [this discussion](https://github.com/proximax-storage/cpp-xpx-chain-sdk/commit/d3ef2762cb67c9db3ce27f2b4ce975472d7c856b#diff-b17e2305ea701bc695ebdd1f49cb657bR366) with @pananton13

**Review by pananton — COMMENTED** [review]
* Introducing conan is great. Please remove also ./cmake/CMakeLists-rapidjson.txt.in file as it is not used anymore.
* I think that it will be better to fix bug in json parser in a way I've mentioned in a [comment](https://github.com/proximax-storage/cpp-xpx-chain-sdk/pull/1#commitcomment-34153125) regarding this commit.
* Also I'm still against recursive globbing of sources for the reasons described in discussion. Suggest creating a SourceList.cmake file in ./cmake where variables holding sources are initialized and including it to CMakeLists.txt

File: src/sdk/crypto/private_key.cpp [hunk 2]
```diff
@@ -19,7 +19,7 @@ namespace nem2_sdk {
 	
 	PrivateKey::PrivateKey(const supplier<uint8_t>& generator)
 	{
-		std::generate_n(key_.begin(), key_.end(), generator);
+		std::generate_n(key_.begin(), (long int) key_.end(), generator);
```
Comment (pananton) [inline src/sdk/crypto/private_key.cpp:22]: It was a bug here - see ```generate_n``` signature here https://en.cppreference.com/w/cpp/algorithm/generate_n . Second parameter should be number of elements to generate, not the iterator to end. Your fix is incorrect too, please replace to ```key_.end() - key_.begin()```

**Comment by eugene-babichenko** [comment]
@carlocayos Please don't merge it right now, I will also check the build on Windows 10 tomorrow.

**Comment by pananton** [comment]
@eugene-babichenko I'm also stared to fix build on windows (needed to compile project from win pc for myself and found that it does not build actually). There are few errors there, I fix a few and willf ix remaining tommorow.

**Comment by pananton** [comment]
@eugene-babichenko Btw, I will create issues for work that I plan to do myself in order to not intersect with your work.

**Comment by eugene-babichenko** [comment]
@pananton13 I rebased this onto your recent changes. Please check if you are able to build it.

**Review by carlocayos — APPROVED** [review]

**Pull request merged**